airflow/providers/sftp/operators/sftp.py

```diff
@@ -134,29 +175,81 @@ def execute(self, context: Any) -> str:
-                            sftp_client=sftp_client,
-                            remote_directory=remote_folder,
+                if self.local_filepath and self.remote_filepath:
+                    if isinstance(self.local_filepath, list) and isinstance(self.remote_filepath, str):
+                        for file_path in self.local_filepath:
+                            local_folder = os.path.dirname(file_path)
+                            local_file = os.path.basename(file_path)
+                            file_msg = file_path
+                            self._transfer(sftp_client, local_folder, local_file, self.remote_filepath)
+                    elif isinstance(self.remote_filepath, list) and isinstance(self.local_filepath, str):
+                        for file_path in self.remote_filepath:
+                            remote_folder = os.path.dirname(file_path)
+                            remote_file = os.path.basename(file_path)
+                            file_msg = file_path
+                            self._transfer(sftp_client, self.local_filepath, remote_file, remote_folder)
+                    elif isinstance(self.remote_filepath, str) and isinstance(self.local_filepath, str):
+                        local_folder = os.path.dirname(self.local_filepath)
+                        file_msg = self.local_filepath
+                        self._transfer(
+                            sftp_client,
+                            local_folder,
+                            self.local_filepath,
+                            self.remote_filepath,
+                            only_file=True,
                         )
-                    file_msg = f"from {self.local_filepath} to {self.remote_filepath}"
-                    self.log.info("Starting to transfer file %s", file_msg)
-                    sftp_client.put(self.local_filepath, self.remote_filepath, confirm=self.confirm)
+                elif self.local_folder and self.remote_folder:
+                    if self.operation.lower() == SFTPOperation.PUT:
+                        files_list = self._search_files(os.listdir(self.local_folder))
+                        for file in files_list:
+                            local_file = os.path.basename(file)
+                            file_msg = file
+                            self._transfer(sftp_client, self.local_folder, local_file, self.remote_folder)
+                    elif self.operation.lower() == SFTPOperation.GET:
+                        files_list = self._search_files(sftp_client.listdir(self.remote_folder))
+                        for file in files_list:
+                            remote_file = os.path.basename(file)
+                            file_msg = file
+                            self._transfer(sftp_client, self.local_folder, remote_file, self.remote_folder)
+                else:
+                    raise AirflowException(f"Argument mismatch, please read docs \n {SFTPOperator.__doc__}")
```

Comment:
The argument check should be done in `__init__` (i.e. fail on DAG parsing instead of running) and use `TypeError` (the customary exception in Python for argument mismatch). It should also describe more clearly what the mismatch is (e.g. `local_folder cannot be used with remote_file`) instead of including a big block of text.